java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/debugging/attach/AttachConfigurationCompletion.java

```diff
@@ -0,0 +1,176 @@
+            w.name("request").value(AttachConfigurations.CONFIG_REQUEST);   // NOI18N
+            int locationIndex = 2;
+            for (Map.Entry<String, ConfigurationAttribute> entry : configAttrs.getAttributes().entrySet()) {
+                ConfigurationAttribute ca = entry.getValue();
+                if (ca.isMustSpecify()) {
+                    String value = ca.getDefaultValue();
+                    if (value.startsWith("${command:")) { // Do not suggest to customize values provided by commands    // NOI18N
+                        value = Utils.escapeCompletionSnippetSpecialChars(encode2JSON(value));
+                    } else {
+                        value = "${" + (locationIndex++) + (value.isEmpty() ? "}" : ":" + Utils.escapeCompletionSnippetSpecialChars(encode2JSON(value)) + "}"); // NOI18N
+                    }
+                    // We have pre-encoded the value in order not to encode the completion snippet escape characters
+                    w.name(entry.getKey()).jsonValue("\"" + value + "\"");
+                }
+            }
+            w.endObject();
+            w.flush();
+        } catch (IOException ex) {
+            Exceptions.printStackTrace(ex);
+        }
+        ci.setInsertText(sw.toString());
+        ci.setInsertTextFormat(InsertTextFormat.Snippet);
+        ci.setDocumentation(configAttrs.getDescription());
+        return ci;
+    }
+
+    private static String encode2JSON(String value) {
+        if (value.isEmpty()) {
+            return value;
+        }
+        StringWriter sw = new StringWriter();
+        try (JsonWriter w = new JsonWriter(sw)) {
+            w.beginArray();
+            w.value(value);
+            w.endArray();
+            w.flush();
+        } catch (IOException ex) {
+            Exceptions.printStackTrace(ex);
+        }
+        String encoded = sw.toString();
+        // We have ["value"], remove the array and quotes
+        return encoded.substring(2, encoded.length() - 2);
+    }
+
+    private static List<CompletionItem> createAttributesCompletion(AttachConfigurations attachConfigurations, Map<String, Object> currentAttributes) {
+        List<CompletionItem> completionItems = null;
+        ConfigurationAttributes currentConfiguration = attachConfigurations.findConfiguration(currentAttributes);
+        if (currentConfiguration != null) {
+            Map<String, ConfigurationAttribute> attributes = currentConfiguration.getAttributes();
+            for (Map.Entry<String, ConfigurationAttribute> entry : attributes.entrySet()) {
+                String attrName = entry.getKey();
+                if (!currentAttributes.containsKey(attrName)) {
+                    StringWriter sw = new StringWriter();
+                    try (JsonWriter w = new JsonWriter(sw)) {
+                        w.beginObject();
+                        w.name(attrName).value(entry.getValue().getDefaultValue());
```

Comment:
Is escaping unnecessary here ?

Reply:
Yes because `value()` does the escaping itself. `jsonValue()` does not do the escaping.
We need to apply escaping ourselfs and use the `jsonValue()` only when we use `${}` in the snippets so that we do not get it escaped for the snippet apply logic to work.